### PR TITLE
Build: take home solutions

### PR DESCRIPTION
**Pull request by ambreenhasan**

### CODE REVIEW

File: CODE_INSTRUCTIONS.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ The goal is to make incremental improvements and create a learning opportunity;
 
 Another engineer added a few features to this API.
 
-* They updated the `/vagetables` json response to include the average rating for each vegetable. You can find that change in the Vegetable class.
+* They updated the `/vegetables` json response to include the average rating for each vegetable. You can find that change in the Vegetable class.
```
Comment (ambreenhasan) [inline CODE_INSTRUCTIONS.md:11]: Typo

File: app/models/user.rb [hunk 2]
```diff
@@ -4,6 +4,6 @@ class User < ApplicationRecord
   before_save :format_email
 
   def format_email
-    email.downcase!
+    email.downcase! if email
```
Comment (ambreenhasan) [inline app/models/user.rb:7]: Can email ever be nil? It doesn't seem like there are any protections in place that would guarantee email to always have a value. Adding an if-clause protects this from exploding when trying to save an email that has a value of 'nil'. 

In general, it would seem that email would be important to have since one of the few ways to identify a user (if we needed to). In later iterations of this, it would be good to add a database and active model validation on the presence of an email, though if this change has been in production for awhile then we would need to backfill the null email entries before we can add a validation like that.

Comment (ambreenhasan) [reply app/models/user.rb:7]: It would be good to add tests to check for the 'nil' case.

File: app/models/vegetable.rb [hunk 3]
```diff
@@ -3,13 +3,19 @@ class Vegetable < ApplicationRecord
 
   def as_json(options = {})
     number_of_ratings = ratings.count
-    total = 0
+    rating_values = raings.map(&:value)
 
-    ratings.each do |rating|
-      total += rating.value
+     # solution 1:
+    total = rating_values.inject(0) do |running_total, value|
```
Comment (ambreenhasan) [inline app/models/vegetable.rb:9]: Although there isn't anything wrong with the current solution, you could use `inject` here instead since it is more idiomatic Ruby. You won't have to declare a `total` variable, since `inject` allows you to add an accumulator value. 

Also, since we only care about the ratings' value, we could map out just the values so we don't need to iterate over an entire object. This isn't anything major, but I think cleans up the code a bit. 

Adding some testing here would be good too to make sure the logic is sound.

File: db/migrate/20211129190106_add_unique_index_for_email.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddUniqueIndexForEmail < ActiveRecord::Migration[6.0]
+  def change
+    add_index :users, :email, unique: true
```
Comment (ambreenhasan) [inline db/migrate/20211129190106_add_unique_index_for_email.rb:3]: Although we have a model validation on email, it is good to have a database-level validation too since it helps to keep the data consistent (in case it was missed before).

File: db/migrate/20211129191536_add_index_for_vegetables_and_ratings.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddIndexForVegetablesAndRatings < ActiveRecord::Migration[6.0]
+  def change
+    add_index :ratings, :vegetable_id
```
Comment (ambreenhasan) [inline db/migrate/20211129191536_add_index_for_vegetables_and_ratings.rb:3]: It appears we are doing a lot of querying between vegetable and ratings (especially in the `as_json` method in Vegetable class), adding an index would be a good performance benefit especially if we increase the number of requests that use the `as_json` method.
